### ISSUE TITLE
`mixin class` hides super methods of other ascendants.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/sdk#53549]
This is follow up for [previous discussion](https://github.com/dart-lang/sdk/issues/53416)

There is a way to declare `mixin class`. Very popular example is [ChangeNonifier](https://github.com/flutter/flutter/blob/e599dae423d346e1607c9186537386831ccdcc36/packages/flutter/lib/src/foundation/change_notifier.dart#L134) in Flutter Framework.

I experimented with it and could not find a way use `mixin class` so that it does not hide methods with the same name of other super classes. Even if we make sure they the same interface. 
This means, if my class uses ChangeNonifier, it cannot meaningfully use any other disposable mixins or derive from disposable classes. And, it seems to be a problem. 

Did i miss something?

```
import 'package:meta/meta.dart';
import 'package:test/test.dart';

mixin Disposable {
  @mustCallSuper
  void dispose() {
    print('i am Disposable');
  }
}

class Class1 with Disposable {
  @override
  void dispose() {
    print('i am Class1');
    super.dispose();
  }
}

mixin Mixin1 on Disposable {
  @override
  void dispose() {
    print('i am Mixin1');
    super.dispose();
  }
}

mixin class MixinClass1 implements Disposable {
  @override
  @mustCallSuper
  void dispose() {
    print(
        'i am MixinClass1, i will have to implement Disposable, because I cannot extend or be on non object');
  }
}

class SuperPosition1 extends Class1 with Mixin1, MixinClass1 {
  @override
  void dispose() {
    print('i am SuperPosition');
    super.dispose();
  }
}

void main() {
  test('dispose SuperPosition1', () {
    SuperPosition1().dispose();
    // Output shows dispose of Class1 and Mixin1 are missing:
    // i am SuperPosition
    // i am MixinClass1, i will have to implement Disposable, because I cannot extend or be on non Object
  });
}
```

**Comment by lrhn** [comment]
This describes how mixins work.
A mixin applied to a superclass will create a subclass of the superclass. The application adds the mixin's instance members as if they were declared in the subclass, and if they have the same name as a superclass member, they will override the superclass member. And they must be valid overrides.

The problem here is not that behavior, it works precisely as designed.
The problem is how it objects with the annotation `@mustCallSuper`.

If a mixin has a member, like `dispose`, which usually must call super, then the mixin can either call `super.dispose`, in which case it must have an `on` type with a `dispose` method, or out can not call `super.dispose`, in which case it's not safe to mix in on top of something with a dispose that must be called.

There is no way to conditionally call super if it's there.
Which makes it problematic to have a `mixin class` with a `dispose` method, because it cannot call `super.disposr` done it's superclass is `Object`.


Or rather, the way to do that is to have two different types: A mixin with an `on` type and a super-invocation and a base class that you can extend.

```dart
mixin ChangeNotifier on Disposable {
  // ...
  void dispose() {
    // ... do work ...
    super.dispose();
  }
}

class ChangeNotifierBase = Disposable with ChangeNotifier;
```
(Possibly, or something like that, haven't checked whether `Disposable` can be extended.)

Then someone mixing in `ChangeNotifier` will call super, and must do so on an existing disposable, and someone using it as the base superclass must use the class, which adds the superclass `dispose` for it to call.

Not as convenient as today, but what we have today doesn't actually work.

**Comment by lrhn** [comment]
A `mixin class`, by itself, is perfectly reasonable as something that can be used both as a superclass and a mixin.

The problem is the `@mustCallSuper` annotation which interacts particularly with mixins *in general*.

If a mixin has an implementation of a `mustCallSuper` member, it *must* have an `on` supertype with that mixin, and call `super.theMethod()` in its implementation. 
Otherwise it won't be correct to mix it in on top of something which already implements that member. 

That's independent of being a `mixin class`, you'd have the same problem with a plain `mixin` that not have the "Disposable" interface as `on` type, and doesn't call `super.dispose()`.

The `mustCallSuper` can occur in two ways:
* Introducing, the base case which doesn't call `super`, because there is no super to call.
* Incrementing, on top of something which already has the must-call-super method, and which must call `super`.

The same method cannot do both, so each method must be written for one of the two roles.

If you just declare a normal method, you usually know whether you need to call `super`, it depends on the superclass, which you know at the point of declaration.

Mixin methods must also be designed for one of the two roles. The problem here is that `ChangeNotifier` is designed for the "introducing" rule, and therefore it doesn't work for the "incrementing" role.
You can still use the current `ChangeNotifier` to *introduce* a `dispose` method, but it doesn't work as an increment on an existing `dispose` method.

Again, that's entirely specific to `@mustCallSuper` and mixins, it's not about `mixin class`. It's just that a `mixin class` necessarily must have the "introducing" role, since it _cannot_ call super, but it's just as much a problem for a non-`class` mixin with the same method.

**Comment by polina-c** [comment]
I disagree. There is still issue with `mixin class`. 
Normal `mixin` can switch from introducing to incrementing, without static breaking changes, while `mixin class` cannot, because it can be just for Object.

**Comment by lrhn** [comment]
I agree that `mixin class` can have problems with `@mustCallSuper` in general. It can only be used to introduce a `@mustCallSuper` property.
If you *also* need a mixin which adds an increment/propagation of the `@mustCallSuper` member, you'll probably be better off *only* having that mixin.

That is, instead of `base class ChangeNotifierBase` (introducing `dispose`) and `mixin ChangeNotifier` (propagating dispose), only have the latter. Then you'll have to write `extends Disposable with ChangeNotifier` to start disposables at the introducing member.

This is still very much an issue about `@mustCallSuper`. There could be a couple of other annotations with similar role-separations, but most other annotations are probably completely indifferent to whether the superclass has the same annotation or not. So anything we'd warn about should be about `@mustCallSuper`.

I'd suggest warning if *mixing in* a mixin which introduces a `@mustCallSuper` member, if the superclass already has a `@mustCallSuper` member with the same name.
Here "introduces a `@mustCallSuper` member" is defined as having a concrete member the mixin declaration, which is either itself annotated as `@mustCallSuper` or which overrides a superinterface member which was annotated `@mustCallSuper`, but _not_ having such a member in any *superclass* (`on`-type) interface. 

_(That's a property that can be computed for each mixin, once and for all: Which `@mustCallSuper` members does it introduce.)_

Mixing such a mixin on top of a superclass which *does* have a `@mustCallSuper` member with the same name should give a warning of *that mixin application not calling `super`*.

**Comment by polina-c** [comment]
Yes, we need mixin Disposable, that introduces dispose. 
But, in addition we need both class and mixin for ChangeNotifier, that increment dispose, because both already exist if Flutter and people took dependency on them and now we need to provide both of them (they still will be upset because of renaming of one of them).

I agree, both 'mixin class' and @mustCallSuper have issues.

**Comment by polina-c** [comment]
ChangeNotifier was a class initially and in March it was [converted](https://github.com/flutter/flutter/commit/25e38a2a875a7b5edb5e0b7b09143d3b2ae2f43b#diff-b95340b7349d629847db8456501c84f6bd938a4b1ae498be35bf9a285d88d575) to 'mixin class'.

**Comment by lrhn** [comment]
`ChangeNotifier` has not changed behavior, it has always behaved the same way, and had the same problems with `@mustCallSuper`.

Whatever change it's needed to address those problems, either in `ChangeNotifier` itself, or adding a warning when using it unsafely, is probably not so urgent it needs to be rushed.

**Comment by polina-c** [comment]
While, yes, this issue is very old, and, yes, it is not super urgent, it still important, because resources are not released in cases, when users made special effort to release them:

1. It is common case to use ChangeNodifier together with mixin like AutoDispose
2. Mixins like AutoDispose are created  specifically for the goal to dispose resources.
3. If ChangeNodifier is not disposed, it will not release listeners, that can cause memory leaks by keeping the listeners in memory much longer than it is designed.
4. dispose of either ChangeNodifier or AutoDispose' is missed, while users are confident they are both invoked

b/301445398 shows details

I will put together a design doc with suggestion how to address this.


I will create design doc to address this by introducing Disposable and converting ChangeNotifier to pure mixin.

**Comment by polina-c** [comment]
@goderbauer

**Comment by agrapine** [comment]
👀
